--- src/Data/Binary/Get.hs
+++ src/Data/Binary/Get.hs
@@ -267,8 +267,8 @@
                     do let now = B.concat . L.toChunks $ consuming
                        put $! mkState rest (bytes + fromIntegral n)
                        -- forces the next chunk before this one is returned
-                       when (B.length now < n) $
-                         fail "too few bytes"
+                       () <- when (B.length now < n) $
+                               fail "too few bytes"
                        return now
 {-# INLINE getBytes #-}
 -- ^ important
@@ -322,7 +322,9 @@
 
 getPtr :: Storable a => Int -> Get a
 getPtr n = do
-    (fp,o,_) <- readN n B.toForeignPtr
+    (fp,o,l) <- readN n B.toForeignPtr
+    () <- when (l < n) $
+            fail "too few bytes"
     return . B.inlinePerformIO $ withForeignPtr fp $ \p -> peek (castPtr $ p `plusPtr` o)
 {-# INLINE getPtr #-}
 

